daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -370,7 +370,11 @@ private[lf] final class Compiler(
       val tmplId = Identifier(pkgId, QualifiedName(module.name, tmplName))
       addDef(compileCreate(tmplId, tmpl))
       addDef(compileFetchTemplate(tmplId, tmpl))
-      addDef(compileSoftFetchTemplate(tmplId, tmpl, preds))
+      module.definitions.get(tmplName) match {
```

Comment:
Wait this isn't just tmpl? can it be derived from tmpl?

Reply:
that's what I thought! it turns out that `module.templates` contains `GenTemplate`s, which only have the template-specific stuff (`ensure` clause, signatory, choices, etc); the datatype itself is just a `DDataType` `GenDefinition`

https://github.com/digital-asset/daml/blob/08933990726f48ddedc68455fd958b4b179cb7bb/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala#L843-L854

https://github.com/digital-asset/daml/blob/08933990726f48ddedc68455fd958b4b179cb7bb/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala#L669-L676

Reply:
so `tmpl` is a DDataType and `module.definitions.get` gives you the GenTemplate?
or is it the other way around, `module.definitions` implies it wouldn't be template specific.
from that, i'd assume you can derive a DDataType from the GenTemplate, surely?

Reply:
@samuel-williams-da wrote
> Wait this isn't just tmpl? can it be derived from tmpl?

You can see a template definition as a kind of type class instance for a standard record.  
Like in Haskell you have the type definition in one side, and the class instance on the other.